metadata-io/src/main/java/com/linkedin/metadata/timeseries/elastic/ElasticSearchTimeseriesAspectService.java

```diff
@@ -175,4 +186,56 @@ public GenericTable getAggregatedStats(@Nonnull String entityName, @Nonnull Stri
       @Nullable GroupingBucket[] groupingBuckets) {
     return _esAggregatedStatsDAO.getAggregatedStats(entityName, aspectName, aggregationSpecs, filter, groupingBuckets);
   }
+
+  /**
+   * A generic delete by filter API which uses elasticsearch's deleteByQuery.
+   * NOTE: There is no need for the client to explicitly walk each scroll page with this approach. Elastic will synchronously
+   * delete all of the documents matching the query that is specified by the filter, and internally handles the batching logic
+   * by the scroll page size specified(i.e. the DEFAULT_LIMIT value of 10,000).
+   * @param entityName the name of the entity.
+   * @param aspectName the name of the aspect.
+   * @param filter the filter to be used for deletion of the documents on the index.
+   * @return the numer of documents returned.
+   */
+  @Nonnull
+  @Override
+  public DeleteAspectValuesResult deleteAspectValues(@Nonnull String entityName, @Nonnull String aspectName,
+      @Nonnull Filter filter) {
+    final String indexName = _indexConvention.getTimeseriesAspectIndexName(entityName, aspectName);
+    final BoolQueryBuilder filterQueryBuilder = QueryBuilders.boolQuery().must(ESUtils.buildFilterQuery(filter));
+    final DeleteByQueryRequest deleteByQueryRequest = new DeleteByQueryRequest(indexName).setQuery(filterQueryBuilder)
+        .setBatchSize(DEFAULT_LIMIT)
+        .setRefresh(true)
+        .setTimeout(TimeValue.timeValueMinutes(10));
+    try {
+      final BulkByScrollResponse response = _searchClient.deleteByQuery(deleteByQueryRequest, RequestOptions.DEFAULT);
+      return new DeleteAspectValuesResult().setNumDocsDeleted(response.getDeleted());
+    } catch (IOException e) {
+      log.error("Delete query failed:" + e.getMessage());
+      throw new ESQueryException("Delete query failed:", e);
+    }
+  }
+
+  @Nonnull
+  @Override
+  public DeleteAspectValuesResult rollbackTimeseriesAspects(@Nonnull String runId) {
+    DeleteAspectValuesResult rollbackResult = new DeleteAspectValuesResult();
+    // Construct the runId filter for deletion.
+    Criterion hasRunIdCriterion = new Criterion().setField("runId").setCondition(Condition.EQUAL).setValue(runId);
+    Filter filter = QueryUtils.getFilterFromCriteria(ImmutableList.of(hasRunIdCriterion));
+
+    // Delete the timeseries aspects across all entities with the runId.
+    for (Map.Entry<String, EntitySpec> entry : _entityRegistry.getEntitySpecs().entrySet()) {
```

Comment:
Discussion point. Can we somehow avoid going through all timeseries aspect indices when we rollback? Is this worth optimizing for? 

For example for versioned aspect rollbacks, the system metadata services stores the list of aspects the run ingested. Here we are not storing that and instead blindly deleting everything with runid X on all indices

Reply:
This approach is far more efficient than going through the system metadata.